### ISSUE TITLE
Undefined offsets

### ISSUE THREAD

**Issue by dev-101** [Maikuolan/SFS-Mass-IP-Checker#2]
Hi, very useful tool :)

However, while running my first test with some 240+ IPs (some of which were in local db, but not all), 20 notices were generated (5 cycles x 4):

> Notice: Undefined offset: 2 in sfs\SFS-Mass-IP-Checker.php on line 554
> Notice: Undefined offset: 2 in sfs\SFS-Mass-IP-Checker.php on line 569
> Notice: Undefined offset: 1 in sfs\SFS-Mass-IP-Checker.php on line 585
> Notice: Undefined offset: 2 in sfs\SFS-Mass-IP-Checker.php on line 589

Not sure if this is because of SFS limit, some timeout, but thought should report it.
Tested on 2 different servers, same issue.

**testing IPv4 set:**

180.183.139.74
84.46.250.208
187.64.183.111
162.244.8.169
162.244.8.169
162.244.8.169
162.244.8.169
162.244.8.169
130.105.214.137
182.203.243.227
93.127.136.177
39.59.49.237
125.27.42.56
221.193.116.108
110.244.244.192
185.119.252.119
119.7.80.72
110.244.244.192
1.172.89.245
42.113.165.224
42.113.165.224
222.85.102.254
212.66.34.157
5.222.143.220
175.232.44.193
118.137.177.15
93.84.195.152
190.128.255.154
117.169.4.130
183.46.146.177
185.14.193.136
61.31.133.42
196.196.88.251
110.77.221.140
180.250.182.51
196.196.88.224
196.196.94.44
123.96.179.110
103.194.170.54
158.222.9.179
91.126.61.41
80.95.29.212
39.40.203.248
165.231.96.176
115.230.222.102
93.127.147.246
140.240.150.204
139.0.81.237
101.109.184.194
158.222.13.106
187.160.204.213
154.66.246.97
118.172.233.210
111.94.190.125
196.196.94.173
196.196.88.250
162.244.8.169
162.244.8.169
185.139.236.161
118.165.66.91
91.108.72.118
201.173.81.184
201.166.193.99
108.175.11.225
95.188.244.179
39.59.17.112
187.161.117.121
178.155.5.226
95.67.69.26
42.113.167.143
86.164.161.23
42.113.167.143
165.231.96.176
109.147.1.187
37.105.149.58
42.113.167.143
189.219.94.114
121.140.18.186
196.196.88.21
86.130.121.67
162.244.8.164
162.244.8.164
162.244.8.164
162.244.8.164
118.169.118.181
36.73.198.4
165.231.105.15
183.82.109.189
220.141.111.249
110.244.246.113
89.117.242.108
110.244.246.113
123.18.202.176
110.244.246.113
50.190.113.28
42.113.163.232
79.180.246.148
183.83.51.218
106.8.219.16
168.235.93.190
110.244.247.25
14.148.151.59
77.78.91.13
110.244.247.25
37.106.138.155
177.104.206.34
110.244.247.25
180.180.102.211
110.244.247.25
110.244.247.25
180.252.167.53
115.214.232.17
118.137.177.47
46.143.156.20
110.252.66.5
125.27.82.165
106.8.223.216
110.252.66.5
110.252.66.5
213.43.160.206
78.108.108.9
36.235.6.31
118.137.177.47
110.252.66.5
109.163.167.87
182.53.199.127
114.36.72.192
183.7.94.137
143.137.62.2
158.222.14.84
112.226.112.103
120.10.173.44
86.38.182.196
101.30.193.27
110.244.218.189
165.231.99.95
110.244.218.189
183.60.201.158
196.196.90.15
110.0.119.215
222.174.72.122
46.52.220.6
185.104.217.183
84.46.248.34
189.219.173.70
202.84.70.3
106.8.208.28
181.214.37.34
177.205.6.42
185.71.1.69
185.71.2.225
176.193.144.4
119.251.161.161
119.251.161.161
195.228.156.102
119.251.161.161
119.251.161.161
114.113.220.99
180.148.210.180
165.231.99.163
36.81.32.11
165.231.96.221
203.112.72.178
14.67.133.20
196.196.94.173
189.83.252.147
110.244.228.100
162.244.10.84
162.244.8.164
196.196.92.212
110.244.228.100
110.244.228.100
41.230.246.176
110.244.220.163
94.179.214.99
110.244.220.163
110.244.220.163
162.244.8.169
162.244.8.169
162.244.8.169
162.244.8.169
162.244.8.169
162.244.8.169
67.197.139.199
61.55.110.130
114.245.206.68
68.234.139.73
123.112.68.84
86.122.39.145
147.75.203.139
110.244.204.147
180.180.111.168
158.222.8.59
24.44.54.177
188.0.67.24
114.247.108.156
165.231.96.239
45.125.69.243
120.10.240.111
178.94.228.149
120.10.240.111
196.196.88.63
120.10.240.111
27.159.56.198
178.155.5.39
104.254.212.183
114.26.120.65
109.147.140.181
42.113.164.246
86.38.190.76
1.20.184.244
62.210.80.28
106.75.17.129
196.196.88.117
114.26.11.2
110.244.192.93
1.164.151.46
158.222.12.17
110.244.192.93
162.219.172.21
43.245.86.27
196.196.94.173
5.107.47.194
110.138.209.254
192.118.76.147
27.111.30.18
82.199.54.207
36.224.253.54
5.107.47.194
91.108.73.122
46.99.8.244
125.164.11.208
62.210.111.76
103.250.58.52
158.222.12.17
123.77.218.17
196.196.88.51
191.222.106.244
62.210.80.28
183.89.88.9
165.255.164.166
158.222.8.198

**Comment by DanielRuf** [comment]
Which IPs triggered these notices?

**Comment by Maikuolan** [comment]
Hm, interesting. I didn't encounter this problem during the previous most recent round of testing that I'd done (just prior to the latest commits). Could be something new, or.. I might've just missed this problem, I guess.

Anyhow, I've successfully replicated the problem on my end just now (re-tested everything), so, marking this as a bug.

```
2016/09/08 14:13:26 [error] 6120#7808: *26 FastCGI sent in stderr: "PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 554
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 569
PHP Notice:  Undefined offset: 1 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 585
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 589
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 554
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 569
PHP Notice:  Undefined offset: 1 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 585
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 589
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 554
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 569
PHP Notice:  Undefined offset: 1 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 585
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 589
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 554
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 569
PHP Notice:  Undefined offset: 1 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 585
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 589
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 554
PHP Notice:  Undefined offset: 2 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker.php on line 569
PHP Notice:  Undefined offset: 1 in X:\SFS-Mass-IP-Checker\SFS-Mass-IP-Checker
```

I'll take a look at it shortly to try to figure out what the cause is.

Additionally, this occurred on my end when trying to run the above IPs through the checker:

![bugsfs](https://cloud.githubusercontent.com/assets/12571108/18338982/23b071d8-75cf-11e6-82a8-64c5e19748b8.png)

..Which, definitely should not be happening. The two problems may or may not be related; I'll figure that out when I look at it shortly (not sure yet). In any case though, discovered upon attempting to check the aforementioned IPs, so, related at least tangentially, anyhow.

Thanks heaps for reporting this. :-)

I'll reply again after I've taken a closer look at the problem.

**Comment by dev-101** [comment]
Hi Maikuolan & DanielRuf,

Yes, that is exactly what I've got running initial tests and probably at the same spot (near the end of sorted IP range), but later could not reproduce it, even after clearing cache and completely disabling it by removing fn-s, and testing new IPs.

I have modified original code to check results array from SFS:

```
    try {
        $SFSMassIPChecker['Results'] =
            SFSMassIPCheckerCheckIP($SFSMassIPChecker['IPAddr']);
            $f = dirname(__FILE__).'/sfsresults.txt';
            $c = var_export($SFSMassIPChecker['Results'], true);
            file_put_contents($f,$c);
    } catch (\Exception $e) {
        $SFSMassIPChecker['Error'] = $e->getMessage();
    }
```

But, that part seems to be fine and so far could not reproduce the thing.

This can be resolved with isset(), simply avoiding these lines when they happen, but more interesting now is to see what and why that happens. Maybe some of the results don't get stored properly into array which triggers this behavior.

**Comment by Maikuolan** [comment]
_Partially_ fixed; Offset errors shouldn't appear anymore, nor should the weird empty space seen in the above screenshot.

Not _entirely_ fixed yet though; The IPs that were failing to be checked, still aren't being checked for some reason. Digging further; Hopefully should have it fully fixed soon.

**Comment by Maikuolan** [comment]
Took me a while, but I've figured out what's causing it.. Kind of annoyed at myself for not spotting the problem earlier. '-.-

Anyhow, working on a fix now; Should have it committed soon.

**Comment by Maikuolan** [comment]
..And committed. :-)

It _should_ be working correctly now. Mind giving it a quick test using the latest codebase to confirm this? Cheers.

**Comment by dev-101** [comment]
Hi, the problem is I cannot re-trigger the notices again with old version. Not sure why, really... as explained in previous post.

Anyway, I have looked at some changes:

> function SFSMassIPCheckerSortIPv4()

comment: you might consider using natsort() instead [http://php.net/manual/en/function.natsort.php]

Additionally, you can verify IP with simple filter:

```
<?php
$ip = "127.0.0.1";

if (!filter_var($ip, FILTER_VALIDATE_IP) === false) {
    // valid
}
?>
```

And lastly, we can exclude private and reserved ranges, just for prevention (but this is not urgent):

```
if(!filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_NO_PRIV_RANGE | FILTER_FLAG_NO_RES_RANGE) === false) {
    // valid
}
```

Thanks!
Cheers

**Comment by Maikuolan** [comment]
> Hi, the problem is I cannot re-trigger the notices again with old version. Not sure why, really... as explained in previous post.

No worries. I guess I can mark this as fixed (should be fixed now in theory, anyhow, and I can't replicate it anymore on my end either, since the latest commits); If I'm wrong about it being fixed, we'll probably find out about this at some stage or another anyhow. :-)

> comment: you might consider using natsort() instead [http://php.net/manual/en/function.natsort.php]
> 
> _Under limited testing, natsort() appears to work well for IP addresses. For my needs, it is far less code than the ip2long()/long2ip() conversion I was using before._

That could be useful. I haven't really played around much with natsort() in the past, due to that the documentation didn't really explain the way it ordered things in a way that made much sense to me, and I didn't have much need for it before. I'll test it out with IPs though, to see if it's viable for our purposes and etc; Good idea if it works as intended.

Using filter_var() to validate IPs and to check for private ranges sounds like good ideas, too. :-)

_Marking as fixed, but keeping the issue open for a while to address some of the additional suggestions made._

**Comment by Maikuolan** [comment]
Done. :-)

**Comment by dev-101** [comment]
Great :+1:
